File: dsa/week(9)/queueone.py
class QueueUsingArray:
    def __init__(self, size):
        self.__data = [None] * size
        self.__nextIndex = 0
        self.__firstIndex = -1
        self.__size = 0
        self.__capacity = size

    def getSize(self):
        return self.__size

    def isEmpty(self):
        return self.__size == 0

    def enqueue(self, element):
        if self.__size == self.__capacity:
            print("Queue is full")
            return
        self.__data[self.__nextIndex] = element
        self.__nextIndex = (self.__nextIndex + 1) % self.__capacity
        if self.__firstIndex == -1:
            self.__firstIndex = 0
        self.__size += 1

    def front(self):
        if self.isEmpty():
            print("Queue is empty!")
            return None
        return self.__data[self.__firstIndex]

    def dequeue(self):
        if self.isEmpty():
            print("Queue is empty!")
            return None
        ans = self.__data[self.__firstIndex]
        self.__data[self.__firstIndex] = None  
        self.__firstIndex = (self.__firstIndex + 1) % self.__capacity
        self.__size -= 1

        if self.__size == 0:
            self.__firstIndex = -1
            self.__nextIndex = 0

        return ans



if __name__ == "__main__":
    q1 = QueueUsingArray(5)
    print(q1.dequeue()) 
    q1.enqueue(1)
    print(q1.getSize()) 
    q1.enqueue(2)
    print(q1.front()) 
    print(q1.dequeue()) 
    print(q1.getSize()) 

File: dsa/week(9)/queueone.js
class QueueUsingArray {
    #data;
    #nextIndex;
    #size;
    #capacity;
    #firstIndex;
    constructor(s) {
        this.#data = new Array(s).fill(0);
        this.#nextIndex = 0;
        this.#size = 0;
        this.#capacity = s;
        this.#firstIndex = -1; 
    }
    getSize() {
        return this.#size;
    }
    isEmpty() {
        return this.#size === 0;
    }
    enqueue(element) {
        if (this.#size === this.#capacity) {
            console.log('Queue is full');
            return;
        }
        this.#data[this.#nextIndex] = element;
        this.#nextIndex = (this.#nextIndex + 1) % this.#capacity;

        if (this.#firstIndex === -1) {
            this.#firstIndex = 0;
        }
        this.#size++;
    }

    front() {
        if (this.isEmpty()) {
            console.log('Queue is empty!');
            return null;
        }
        return this.#data[this.#firstIndex];
    }
    dequeue() {
        if (this.isEmpty()) {
            console.log('Queue is empty!');
            return null;
        }
        let ans = this.#data[this.#firstIndex];
        this.#firstIndex = (this.#firstIndex + 1) % this.#capacity;
        this.#size--;

        if (this.#size === 0) {
            this.#firstIndex = -1;
            this.#nextIndex = 0;
        }
        return ans;
    }
}


const q = new QueueUsingArray(2);
q.dequeue(); 
console.log(q.isEmpty()); 
q.enqueue(1);
q.enqueue(2);
q.enqueue(3);
// console.log(q.Capacity); 
